src/main/java/com/squareup/auto/value/redacted/AutoValueRedactedExtension.java

```diff
@@ -0,0 +1,129 @@
+
+  @Override
+  public boolean applicable(Context context) {
+    Map<String, ExecutableElement> properties = context.properties();
+    for (ExecutableElement element : properties.values()) {
+      if (isRedacted(element)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  @Override
+  public boolean mustBeAtEnd(Context context) {
+    return false;
+  }
+
+  @Override
+  public String generateClass(Context context, String className, String classToExtend,
+      boolean isFinal) {
+    Name superName = context.autoValueClass().getSimpleName();
+    Map<String, ExecutableElement> properties = context.properties();
+
+    TypeSpec subclass = TypeSpec.classBuilder(className) //
+        .addModifiers(isFinal ? Modifier.FINAL : Modifier.ABSTRACT) //
+        .superclass(TypeVariableName.get(classToExtend)) //
+        .addMethod(generateConstructor(properties)) //
+        .addMethod(generateToString(superName, properties)) //
+        .build();
+
+    JavaFile javaFile = JavaFile.builder(context.packageName(), subclass).build();
+    return javaFile.toString();
+  }
+
+  private MethodSpec generateConstructor(Map<String, ExecutableElement> properties) {
+    List<ParameterSpec> params = Lists.newArrayList();
+    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
+      TypeName typeName = TypeName.get(entry.getValue().getReturnType());
+      params.add(ParameterSpec.builder(typeName, entry.getKey()).build());
+    }
+
+    StringBuilder body = new StringBuilder("super(");
+    for (int i = properties.size(); i > 0; i--) {
+      body.append("$N");
+      if (i > 1) body.append(", ");
+    }
+    body.append(")");
+
+    return MethodSpec.constructorBuilder() //
+        .addParameters(params) //
+        .addStatement(body.toString(), properties.keySet().toArray()) //
+        .build();
+  }
+
+  private MethodSpec generateToString(Name superName, Map<String, ExecutableElement> properties) {
+    MethodSpec.Builder builder = MethodSpec.methodBuilder("toString") //
+        .addAnnotation(Override.class) //
+        .addModifiers(Modifier.PUBLIC) //
+        .returns(ClassName.get(String.class)) //
+        .addCode("return \"$L{\"\n", superName) //
+        .addCode("$>$>");
+
+    int count = 0;
+    for (Map.Entry<String, ExecutableElement> entry : properties.entrySet()) {
+      String propertyName = entry.getKey();
+      ExecutableElement propertyElement = entry.getValue();
+      TypeName propertyType = TypeName.get(entry.getValue().getReturnType());
+
+      builder.addCode("+ \"$N=\" + ", propertyName);
+      if (isRedacted(propertyElement)) {
+        if (propertyType.isPrimitive()) {
+          builder.addCode("\"██\"");
+        } else {
+          builder.addCode("$N() != null ? \"██\" : null", propertyName);
+        }
+      } else if (propertyType instanceof ArrayTypeName) {
+        builder.addCode("$T.toString($N())", Arrays.class, propertyName);
```

Comment:
In Java 10 arrays will have a common supertype (e.g., `Array<int>` for `int[]`) but they said that they can't make `toString()` just work because of backwards compatibility reasons.